MCH: introduce digit modifier in filtering workflow

aferrero2707 opened the pull request:
The digit modifier allows to change the contents of the digits in the filtering step.

It is introduced in order to correct some mapping issues in the CTFs already collected, but the interface is general, and in the future it might be used for any kind of digit manipulation, if needed.

The code also introduces pad remapping functions for some regions of ST1 and ST2 quadrants, where the mapping used until the 2024 data taking was wrong. The digits modifier uses the ST1 and ST2 remapping functions in order to fix the mapping of the already existing CTFs.

The ST1 and ST2 mappings are fixed by https://github.com/AliceO2Group/AliceO2/pull/13996 and https://github.com/AliceO2Group/AliceO2/pull/13938.

More details can be found in https://its.cern.ch/jira/browse/MCH-4 and https://its.cern.ch/jira/browse/MCH-5.

aferrero2707 commented:
@pillot @lmassacr those are the fixes to the ST2 mapping that were discussed recently. As for https://github.com/AliceO2Group/AliceO2/pull/13921, we need to look closely to the implications for the reading of CCDB calibration objects - like the BadChannels.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitFilteringSpec.cxx:
```diff
@@ -126,6 +142,9 @@ class DigitFilteringTask
             }
+
+            // modify the digit if needed
+            mDigitModifier(oDigits.back());
```
I am wondering if this modification should be done before or after checking the goodness of the digit? In particular w.r.t the statusMap. Is the bug in the mapping also affect the CCDB objects like BadChannels? And/or their conversion to padIds to compare with the padId of the digit?

aferrero2707 replied:
You are right, it should be done before, such that the correct digit mapping is used when comparing with the StatusMap. The only caveat is that, as far as I can see, it requires to always create a local copy of the digit.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitFilteringSpec.cxx:
```diff
@@ -160,14 +179,18 @@ class DigitFilteringTask
   int32_t mTimeCalib{0};
   uint32_t mStatusMask{0};
+  bool mCorrectST1Mapping;
+  bool mCorrectST2Mapping;
```
I think it's better to explicitly initialize the data members (to false here).

aferrero2707 replied:
Indeed.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitModifier.cxx:
```diff
@@ -0,0 +1,142 @@
+        int channelForPadIdMin = -1;
+        for (int channel = 0; channel < 64; channel++) {
+          auto padId = segment.findPadByFEE(dsId, int(channel));
+          if (padIdMin < 0 || padId < padIdMin) {
```
You do not check the validity of the padId. I suppose it's ok to skip this check if all of the 64 channels are connected for these DS so you are sure they are all valid?
Also this implementation of the remapping assume a certain implementation of the mapping, where all the channels get consecutive indices. Is it the case for the Impl4? just to be sure it is not an implicit assumption.

aferrero2707 replied:
Good point, I will add a check nevertheless. Since this is not expected because all channels are connected in the region we want to fix, maybe an exception would be appropriate?

pillot replied:
Given that if that is not true the entire logic fall apart, an exception makes sense to me yes.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitModifier.cxx:
```diff
@@ -0,0 +1,142 @@
+      // check if the current DE needs some remapping
+      if (padsRemapping.count(digit.getDetID()) > 0) {
+        // check if the current padID needs to be remapped
+        auto& padsRemappingForDe = padsRemapping[digit.getDetID()];
```
It's a micro improvement, but I would use the method find() to test and use the returned iterator instead of count() and operator[], to have 1 single search of the key into the map instead of 2.

aferrero2707 replied:
Ok.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitModifier.cxx:
```diff
@@ -0,0 +1,142 @@
+        auto& padsRemappingForDe = padsRemapping[digit.getDetID()];
+        if (padsRemappingForDe.count(digit.getPadID()) > 0) {
+          // get the corrected padID
+          int padIDRemapped = padsRemappingForDe[digit.getPadID()];
```
same comment here

aferrero2707 replied:
Ok.

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitModifier.cxx:
```diff
@@ -0,0 +1,142 @@
+  return [parts](Digit& digit) {
+    for (const auto& p : parts) {
+      p(digit);
+    }
```
The correctors are mutually exclusive, so I would test the DEId here are call the right one instead of calling the 2. And you can remove the check of the DEId in each of them. Since these functions are hidden to the user one can assume that the internal implementation uses them correctly. And in any case, the remapping tables only have entries for the relevant DE so this check is redundant.

aferrero2707 replied:
Good point.

pillot reviewed (commented):
Thanks @aferrero2707 . Appart from details of implementation, I am also wondering here about the connexion between the digit padId and the BadChannels (and the statusMap in general) and how this is affected by this bug in the mapping and the fix.

aferrero2707 commented:
@pillot I updated the code with all the changes we discussed, could you please have a final look?
The corresponding fix to the mapping is also ready, see https://github.com/AliceO2Group/AliceO2/pull/13938

alibuild commented:
Error while checking build/O2/fullCI_slc9 for e51f1d35d59a7fad9879a738b91554a624175b4a at 2025-02-05 13:05:
```
## sw/BUILD/O2Physics-latest/log
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:
Error in cling::AutoLoadingVisitor::InsertIntoAutoLoadingState:


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile reco_NOGPU.log
reco_NOGPU.log:[27395:internal-dpl-ccdb-backend]: [13:05:11][ERROR] Exception while running: Fatal error. Rethrowing.
reco_NOGPU.log-[27395:internal-dpl-ccdb-backend]: [13:05:11][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-itsmft-stf-decoder-workflow, device shutting down. Reason: Fatal error
[27395:internal-dpl-ccdb-backend]: [13:05:11][ERROR] CCDBDownloader CURL transfer error - Unsupported protocol
[27395:internal-dpl-ccdb-backend]: [13:05:11][ERROR] CcdbDownloader finished transfer http://alice-ccdb.cern.ch/GLO/Param/MatLUT/1546300800000/3f0d0f81-99d9-11ee-b9a2-200114580202 for 1550600800000 (agent_id: alimetal01.cern.ch-1738757103-JLT3TZ) with http code: 0
[27395:internal-dpl-ccdb-backend]: [13:05:11][FATAL] Unable to find object GLO/Param/MatLUT/1550600800000
[27395:internal-dpl-ccdb-backend]: [13:05:11][ERROR] Exception while running: Fatal error. Rethrowing.
[27395:internal-dpl-ccdb-backend]: [13:05:11][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-itsmft-stf-decoder-workflow, device shutting down. Reason: Fatal error
[ERROR] Workflow crashed - PID 27395 (internal-dpl-ccdb-backend) did not exit correctly however it's not clear why. Exit code forced to 128.
[ERROR]  - Device internal-dpl-ccdb-backend: pid 27395 (exit 128)
[INFO]    - First error: [13:05:11][FATAL] Unable to find object GLO/Param/MatLUT/1550600800000
[ERROR] SEVERE: Device internal-dpl-ccdb-backend (27395) had at least one message above severity 5: Unable to find object GLO/Param/MatLUT/1550600800000


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep ' error:'
++ grep -v clang-diagnostic-error error-log.txt
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/3d5937117926b92747c1c62e81d4f3672e6afcb2/slc9_x86-64/o2checkcode/1.0-local285/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13924/e51f1d35d59a7fad9879a738b91554a624175b4a/build_O2_fullCI_slc9/pretty.html).

pillot commented on Detectors/MUON/MCH/DigitFiltering/src/DigitModifier.cxx:
```diff
@@ -0,0 +1,155 @@
+            // this should never occur in this specific case, as all channels of this group of boards
+            // is connected to pads, hence we rise an exception
+            throw std::out_of_range(std::format("Unknown padId for DE{} DS{} channel {}", deId, dsId, channel));
+          }
```
Hi @aferrero2707 , the test on MacOS fails because of std::format. I think C++20 is not yet fully supported and we should still use fmt::format.

aferrero2707 replied:
Thanks, replaced with `fmt::format()`.

aferrero2707 committed:
[MCH] introduce digit modifier in filtering workflow

The digit modifier allows to change the contents
of the digits in the filtering step.

It is introduced in order to correct some mapping
issues in the CTFs already collected, but the
interface is general, and in the future it might be
used for any kind of digit manipulation, if needed.

aferrero2707 committed:
[MCH] added pad remapping function for ST1

The remapping function corrects the pads mapping
for seven DS boards on the edge of the bending
cathodes, all corresponding to the motif type "1BG".

aferrero2707 committed:
[MCH] added pad remapping function for ST2

The remapping function corrects the pads mapping
for five consecutive motif types on the bending
side of ST2 quadrants (types "2Bv1" to "2Bv5").

aferrero2707 commented:
@pillot I have added the remapping for ST1, and refactored the code to better check if a given pad is a candidate for remapping, to minimize the number of accesses to the unordered_maps.

pillot reviewed (dismissed):
Hi @aferrero2707 , thanks for the new fixes! Looks fine to me.

aferrero2707 commented:
+async-label async-2024-PbPb-apass2 async-2023-PbPb-apass5

pillot commented:
@aferrero2707 , should I squash the commits or let them separated ?

aferrero2707 commented:
> @aferrero2707 , should I squash the commits or let them separated ?

The can be squashed, it was mainly intended to simplify the review.
